python/lsst/obs/base/ingest.py

```diff
@@ -318,16 +365,181 @@ def _calculate_dataset_info(self, header, filename):
+                    self.log.warning("Problem reading index from %s, skipping it: '%s'", path, e)
+                    bad_index_files.add(path)
+                    continue
+
+                self.log.debug("Extracted index metadata from file %s", path)
+
+                # Record that we used this index file
+                good_index_files.add(path)
+
+                for relfile, metadata in index.items():
+                    file = os.path.abspath(os.path.join(directory, relfile))
+                    if file in index_entries:
+                        # ObservationInfo overrides raw metadata
+                        if isinstance(metadata, ObservationInfo) \
+                                and not isinstance(index_entries[file], ObservationInfo):
+                            self.log.warning("File %s already specified in an index file but overriding"
+                                             " with ObservationInfo content from %s", file, path)
+                        else:
+                            self.log.warning("File %s already specified in an index file, "
+                                             "ignoring content from %s", file, path)
+                    else:
+                        index_entries[file] = metadata
+                continue
+
+            by_directory.setdefault(directory, set()).add(file_in_dir)
+
+        # Look for index files in those directories
+        for directory, files_in_directory in by_directory.items():
+            possible_index = os.path.join(directory, "_index.json")
+            if os.path.exists(possible_index):
+                # if a derived index file is found we should only use
+                # information for requested files.
+                try:
+                    index = read_index(possible_index, force_dict=True)
+                except Exception as e:
+                    if self.config.failFast:
+                        raise RuntimeError("Problem reading index from inferred "
+                                           f"location {file_in_dir}") from e
+                    self.log.warning("Problem reading index from inferred location %s, skipping it: %s",
+                                     file_in_dir, e)
+                    bad_index_files.add(possible_index)
+                else:
+                    good_index_files.add(possible_index)
+                    self.log.debug("Extracted potential index metadata from inferred file %s", possible_index)
+
+                for file_in_dir in files_in_directory:
+                    if file_in_dir in index:
+                        file = os.path.abspath(os.path.join(directory, file_in_dir))
+                        if file in index_entries:
+                            # ObservationInfo overrides raw metadata
+                            if isinstance(index[file_in_dir], ObservationInfo) \
+                                    and not isinstance(index_entries[file], ObservationInfo):
```

Comment:
Is this `if` block of warnings almost the same as the above? Could something here be split off into a function or closure?

Reply:
It's all been combined into a single case.